src/async_writer.cpp

```diff
@@ -32,18 +33,40 @@ void AsyncWriter::setConnectionAliveHandler(OnConnectionAliveCallback connection
     this->connectionAliveCallback = connectionAliveCallback;
 }
 
-void AsyncWriter::queueWriteTasks(std::vector<WriteTask>&& tasks)
+void AsyncWriter::setWriteTimedOutHandler(OnWriteTaskTimedOutCallback writeTaskTimeoutHandler)
+{
+    this->writeTaskTimeoutHandler = writeTaskTimeoutHandler;
+}
+
+void AsyncWriter::queueBatchWrite(BatchedWriteTasks&& tasks, OptionalWriteDeadline&& deadlineTime)
 {
     bool writing = !writeTasksQueue.empty();
-    writeTasksQueue.push(std::move(tasks));
+    writeTasksQueue.push(std::pair(std::move(tasks), std::move(deadlineTime)));
     if (!writing)
     {
         doWrite(writeTasksQueue.front());
     }
 }
 
-void AsyncWriter::doWrite(const std::vector<WriteTask>& tasks)
+void AsyncWriter::doWrite(const BatchedWriteTasksWithDeadline& tasksWithDeadline)
 {
+    const auto& tasks = tasksWithDeadline.first;
+    const auto& deadlineTimepoint = tasksWithDeadline.second;
+
+    if (deadlineTimepoint.has_value())
+    {
+        auto now = std::chrono::system_clock::now();
```

Comment:
A more appropriate choice would be `std::steady_clock` or `std::high_performance_clock`. `std::system_clock` is NOT a monotonically increasing clock, and it can have a gap or an overlap.